### PR TITLE
Fulltext search queries

### PR DESCRIPTION
**Pull request by fordN**
See the [Fulltext Search RFC](https://github.com/graphprotocol/rfcs/pull/11) for a detailed description of the feature.

Dependent on [graph-cli PR#433](https://github.com/graphprotocol/graph-cli/pull/433)

----

###### Adding a fulltext search query to the subgraph
```graphql
type _Schema_ 
  @fulltext(
    name: "bandSearch",
    language: en,
    algorithm: rank,
    include: [
      {
        entity: "Band",
        fields: [
          { name: "name" },
          { name: "description" },
          { name: "bio" },
        ]
      }      
    ]
  ) 

type Band @entity {
    id: ID!
    name: String!
    description: String!
    bio: String
    wallet: Address
    labels: [Label!]!
    discography: [Album!]!
    members: [Musician!]!
}
```

###### Using the fulltext search query
```graphql
query {
  bandSearch(text: "breaks & electro & detroit") {
    id
    name 
    description
    wallet
  }
}
```

### CODE REVIEW

File: store/postgres/src/relational_queries.rs [hunk 1]
```diff
@@ -213,6 +219,24 @@ impl<'a> QueryFragment<Pg> for QueryValue<'a> {
                         out.push_sql("[]");
                         Ok(())
                     }
+                    // TSVector will only be in a Value::List() for inserts so "to_tsvector" can always be used here
+                    ColumnType::TSVector((language, _)) => {
+                        let mut iter = values.iter().peekable();
+                        out.push_sql("(");
+                        while let Some(value) = iter.next() {
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs]: Everywhere else in the file where we need to generate a list of things with a delimiter in between, we use the idiom
```rust
   for (i, thing) in things.iter().enumerate() {
     if i > 0 {
         // output delimiter
     }
     // output thing
   }
```

This code should follow that pattern.

File: store/postgres/src/relational_queries.rs [hunk 2]
```diff
@@ -213,6 +219,24 @@ impl<'a> QueryFragment<Pg> for QueryValue<'a> {
...
+                            out.push_sql("to_tsvector(");
+                            out.push_bind_param::<Text, _>(&language.as_sql())?;
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs]: Since we know that `language` is a safe string, you could just write it out directly without going through a bind variable.

File: store/postgres/src/relational_queries.rs [hunk 3]
```diff
@@ -532,7 +561,9 @@ impl<'a> QueryFilter<'a> {
     ) -> QueryResult<()> {
         let column = self.column(attribute);
 
-        if column.is_text() && value.is_string() {
+        if column.is_fulltext() {
+            PrefixComparison::new(Comparison::Match, column, value).walk_ast(out.reborrow())?;
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs]: I don't think we want to go through the `PrefixComparison` machinery here which is just a workaround for the fact that we allow arbitrarily large strings in `text` columns, but only index a prefix of them. It should be possible to turn fulltext searches directly into something like `<column> @@ <value>` like we do for other types.

File: store/postgres/src/relational_queries.rs [hunk 4]
```diff
@@ -1250,17 +1281,19 @@ impl<'a> FilterCollection<'a> {
 /// is `None`, the sort key should be ignored
 #[derive(Debug, Clone)]
 pub struct SortKey {
-    name: Option<SqlName>,
+    column: Option<Column>,
+    value: Option<Value>,
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs]: The code in this file is very careful to do a minimal amount of allocations; I was a little lazy with the `name` (though it was only a string) Now that we are storing more complex data here, we should use references, i.e., this should become
```rust
pub struct SortKey<'a> {
  column: Option<&'a Column>,
  value: Option<&'a Value>
}
```

File: store/postgres/src/relational_queries.rs [hunk 5]
```diff
@@ -1270,16 +1303,43 @@ impl SortKey {
...
-                out.push_sql(", ");
-                out.push_identifier(PRIMARY_KEY_COLUMN)?;
+        if let Some(column) = &self.column {
+            match &column.column_type {
+                ColumnType::TSVector((_, algorithm)) => {
+                    let sort_value = &self
+                        .value
+                        .as_ref()
+                        .and_then(|v| v.clone().as_string())
+                        .expect("fulltext search queries can only use EntityFilter::Equal");
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs]: Generating errors during SQL generation (this function is ultimately called from `walk_ast` somewhere) leads to really confusing and hard to follow error messages because of how things bubble up in Diesel. This check should be done when the `SortKey` is constructed, and it should therefore have a `value: Option<String>` or (preferrably) `value: Option<&str>` field.

File: graph/src/data/subgraph/schema.rs [hunk 6]
```diff
@@ -70,6 +70,12 @@ pub trait TypedEntity {
     }
 }
 
+// Maps an entity to its fields that are covered by a Fulltext API.
+pub type SubgraphFulltextEntities = BTreeMap<Name, SubgraphFulltextFields>;
+
+// Maps an entities Fulltext included fields to the Fulltext fields to which they contribute.
+pub type SubgraphFulltextFields = BTreeMap<Name, Vec<Name>>;
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/subgraph/schema.rs]: haha,  I just added the `Subgraph` prefix in the final hour because we seem to be using that convention in the subgraph/schema naming.  No arguments here.

File: graphql/src/store/query.rs [hunk 7]
```diff
@@ -93,12 +93,36 @@ fn build_filter(
 ) -> Result<Option<EntityFilter>, QueryExecutionError> {
     match arguments.get(&"where".to_string()) {
         Some(q::Value::Object(object)) => build_filter_from_object(entity, object),
-        None | Some(q::Value::Null) => Ok(None),
+        Some(q::Value::Null) => Ok(None),
+        None => match arguments.get(&"text".to_string()) {
+            Some(q::Value::Object(filter)) => build_text_filter_from_object(entity, filter),
```
[the remark this answers is not in the document]
Comment (fordN) [reply graphql/src/store/query.rs]: It's submitted as a string in the query, but I then turn it into an object containing both the field name (for the fulltext field) and string value supplied in order to have the required information to build the filter and order by.

File: store/postgres/src/store.rs [hunk 8]
```diff
@@ -1110,6 +1111,15 @@ impl SubgraphDeploymentStore for Store {
         Ok(self.cached_schema(subgraph_id)?.api)
     }
 
+    fn fulltext_fields(
+        &self,
+        subgraph_id: &SubgraphDeploymentId,
+    ) -> Result<SubgraphFulltextEntities, Error> {
+        Ok(Schema::subgraph_fulltext_entity_fields(
```
[the remark this answers is not in the document]
Comment (fordN) [reply store/postgres/src/store.rs]: The naming is tricky here because the result is a nested map of entity name -> that entities fulltext included fields -> the fulltext query field that the included field contributes to. 

You'll see the SubgraphFulltextEntities map has an inner type SubgraphFulltextFields, so ultimately this data is used to map an entity to fulltext fields hence the method name.

File: graphql/src/execution/execution.rs [hunk 9]
```diff
@@ -1055,7 +1055,14 @@ where
         let value = qast::get_argument_value(&field.arguments, &argument_def.name).cloned();
         match coercion::coerce_input_value(value, &argument_def, &resolver, &ctx.variable_values) {
             Ok(Some(value)) => {
-                coerced_values.insert(&argument_def.name, value);
+                if argument_def.name == "text".to_string() {
+                    coerced_values.insert(
+                        &argument_def.name,
+                        q::Value::Object(BTreeMap::from_iter(vec![(field.name.clone(), value)])),
+                    );
```
[the remark this answers is not in the document]
Comment (fordN) [reply graphql/src/execution/execution.rs]: Later on the field name is needed along with the value when building the order by and filter, so here I'm converting the string input to an object with the field name as a key to the value.

File: graph/src/data/store/mod.rs [hunk 10]
```diff
@@ -458,10 +458,39 @@ impl Entity {
         for (key, value) in update.0.into_iter() {
             match value {
                 Value::Null => self.remove(&key),
+                Value::String(s) => self.insert(key, Value::String(s)),
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/store/mod.rs]: Yes! Thanks for catching.

File: graph/src/data/schema.rs [hunk 11]
```diff
@@ -113,21 +114,21 @@ impl TryFrom<&String> for FulltextLanguage {
...
+            "DE" => Ok(FulltextLanguage::German),
+            "HU" => Ok(FulltextLanguage::Hungarian),
+            "IT" => Ok(FulltextLanguage::Italian),
+            "NO" => Ok(FulltextLanguage::Norwegian),
+            "PT" => Ok(FulltextLanguage::Portugese),
+            "RO" => Ok(FulltextLanguage::Romanian),
+            "RU" => Ok(FulltextLanguage::Russian),
+            "ES" => Ok(FulltextLanguage::Spanish),
+            "SV" => Ok(FulltextLanguage::Swedish),
+            "TR" => Ok(FulltextLanguage::Turkish),
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs]: Works for me.  👍

File: graph/src/components/store.rs [hunk 12]
```diff
@@ -1478,13 +1489,19 @@ impl EntityCache {
                     // Merging with an empty entity removes null fields.
                     let mut data = Entity::new();
                     data.merge_remove_null_fields(updates);
+                    if let Some(fulltext_fields) = fulltext_entity_fields.get(&key.entity_type) {
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/components/store.rs]: I don't agree, then we are passing an Entity into the store that is partially updated, but still contains the old value for any fulltext fields.  

I explored how to move this update into `apply_entity_modifications()` or `InsertQuery::new()` , but I don't see how that makes things simpler.  To make the changes in InsertQuery::new I'd need access to a reference of which fields are included in fulltext fields, so that would mean adding a HashSet of fulltext fields to Column or a reference of all fulltext fields to their included fields on the Table.  The other issue is that we are passing in a reference to the Entity that will be applied, would we then need to pass in a mutable reference? It feels much cleaner to me to update the EntityModifications all at the same time, still pass in a reference to the Entity and not require passing fulltext field references down into the query execution.

File: graph/src/data/query/error.rs [hunk 13]
```diff
@@ -57,6 +57,7 @@ pub enum QueryExecutionError {
     UndefinedFragment(String),
     // Using slow and prefetch query resolution yield different results
     IncorrectPrefetchResult { slow: q::Value, prefetch: q::Value },
+    FullTextQueryRequiresEqualFilter,
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/query/error.rs]: Ok..Isn't that the same though?

[the remark this answers is not in the document]
Comment (Jannis) [reply graph/src/data/query/error.rs]: Haha, you're right. What was I thinking? (Well, almost, it should be `Fulltext`, not `FullText`.)

File: store/postgres/src/store.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (fordN) [reply store/postgres/src/store.rs]: This is going away anyway, as I'm moving the fulltext_entities data to the Layout.

File: graph/src/data/schema.rs [hunk 14]
```diff
@@ -606,6 +793,194 @@ impl Schema {
...
+                                    .is_some();
+                                !field_exists
+                            })
+                            .is_some()
+                    }
+                    // No entity matches the entity include object
+                    None => return true,
+                }
+            }
+            _ => true,
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs]: Hmm this section _is_ a bit confusing. I'll rework it.

File: graph/src/data/schema.rs [hunk 15]
```diff
@@ -913,6 +1288,43 @@ impl Schema {
             .into_iter()
             .find(|object_type| object_type.name.eq(SCHEMA_TYPE_NAME))
     }
+
+    pub fn entity_fulltext_definitions<'a>(
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs]: This should only run once: during deploy when generating the Layout for the relational schema.   :)

File: graphql/src/store/query.rs [hunk 16]
```diff
@@ -86,19 +86,43 @@ fn build_range(
...
 fn build_filter(
     entity: ObjectOrInterface,
     arguments: &HashMap<&q::Name, q::Value>,
 ) -> Result<Option<EntityFilter>, QueryExecutionError> {
     match arguments.get(&"where".to_string()) {
         Some(q::Value::Object(object)) => build_filter_from_object(entity, object),
-        None | Some(q::Value::Null) => Ok(None),
+        Some(q::Value::Null) => Ok(None),
+        None => match arguments.get(&"text".to_string()) {
+            Some(q::Value::Object(filter)) => build_fulltext_filter_from_object(entity, filter),
```
[the remark this answers is not in the document]
Comment (fordN) [reply graphql/src/store/query.rs]: Yea. This is where I use that object because the entity here is from the input schema, so it doesn't have the fulltext fields.

File: store/postgres/src/relational.rs [hunk 17]
```diff
@@ -670,12 +677,25 @@ impl Column {
...
+        })
+    }
+
+    fn new_fulltext(def: &FulltextDefinition) -> Result<Column, StoreError> {
+        SqlName::check_valid_identifier(&def.name, "attribute")?;
+        let sql_name = SqlName::from(def.name.as_str());
+
+        Ok(Column {
+            name: sql_name,
+            field: def.name.to_string(),
```
[the remark this answers is not in the document]
Comment (fordN) [reply store/postgres/src/relational.rs]: Adding now :)

File: graph/src/data/schema.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply graph/src/data/schema.rs]: Oh, ok. Nevermind then.

File: graph/src/data/schema.rs [hunk 18]
```diff
@@ -606,6 +815,193 @@ impl Schema {
...
+            _ => return vec![SchemaValidationError::FulltextNameUndefined],
+        };
+
+        let local_types: Vec<&ObjectType> = self
+            .document
+            .get_object_type_definitions()
+            .into_iter()
+            .collect();
+
+        // Validate that the fulltext field doesn't collide with any entity fields
```
Comment (Jannis) [inline graph/src/data/schema.rs]: Thinking about this again: I think we should allow fulltext fields to have the same name as entity fields after all.

From a subgraph developer's perspective, there's no possibility of collisions because the fulltext search field is global (on the `Query` type) and entity fields are local (on the entity type). So we should be allowing overlaps and handle this internally.

My suggestion would be to map fulltext directives named `foo` to fields named `_fulltext_foo`. This has several benefits:

1. No collisions with entity fields possible (assuming we disallow fields with a leading `_`, which we will).
2. It's clear which fields come from which `_Schema_` directive (right now, there is only `fulltext`, but who knows what else we'll add).

Comment (Jannis) [inline graph/src/data/schema.rs]: There is something we should validate instead of entity field collisions though:

For every entity type `Foo`, we generate top-level fields `foo` and `foos`. We have to check for collisions with these. This check should reuse the singular and plural field naming logic from the API schema module.

Comment (fordN) [reply graph/src/data/schema.rs]: I'm not clear on your suggestion here, where would the fields be named like `_fulltext_foo`?
In the places I'm thinking of the fulltext fields are clearly marked, such as fulltext [Columns.](https://github.com/graphprotocol/graph-node/blob/ford/text/store/postgres/src/relational.rs#L734)

File: graph/src/data/schema.rs [hunk 19]
```diff
@@ -73,8 +73,204 @@ pub enum SchemaValidationError {
...
+            directive.argument("algorithm").unwrap().as_enum().unwrap(),
+        )
+        .unwrap();
+
+        let language =
+            FulltextLanguage::try_from(directive.argument("language").unwrap().as_enum().unwrap())
+                .unwrap();
+
+        let included_entity_list = directive.argument("include").unwrap().as_list().unwrap();
+        // Currently fulltext query fields are limited to 1 entity, so we just take the first (and only) included Entity
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs]: Yes, I believe that aside from updates to validation it'd be mostly contained to query side changes.  I'd like to reread our interface implementation and align with @lutter before planning since they will be similar, but to give you an idea of where we'll likely go: 
- We'd still use the same table structure in which each entity table includes an additional field for each fulltext query field.  These fulltext fields are updated as rows in the specific entity table are updated.  For a fulltext field that covers multiple entities, the field well have columns in each of the respective entity tables and at query time the tables will be joined together and the text search vectors (fulltext field data) concatenated together before applying the text search and filtering. We'll need to update the fulltext data that is stored in the Layout to support query generation that joins all relevant entity tables together. 
- The result type for fulltext query fields will be an interface which will be added to the schema on deployment. It will also likely be represented as an entity on the schema, so query generation can proceed similarly to other entity queries.

[the remark this answers is not in the document]
Comment (Jannis) [reply graph/src/data/schema.rs]: Yes, that all matches my expectations. 👍

File: graph/src/data/schema.rs [hunk 20]
```diff
@@ -73,8 +74,204 @@ pub enum SchemaValidationError {
...
+    #[fail(
+        display = "Fulltext 'include' object must contain 'entity' (String) and 'fields' (List) attributes"
+    )]
+    FulltextIncludeEntityMissingOrIncorrectAttributes,
+    #[fail(display = "Fulltext directive includes an entity not found on the subgraph schema")]
+    FulltextIncludedEntityNotFound,
+    #[fail(display = "Fulltext include field must have a 'name' attribute")]
+    FulltextIncludedFieldMissingRequiredProperty,
+    #[fail(
+        display = "Fulltext included Entity field, {}, not found or not a string",
```
Comment (Jannis) [inline graph/src/data/schema.rs]: Perhaps we can shorten this message to `Fulltext entity field `{}` not found or not a string`?

File: graph/src/data/schema.rs (hunk 18 above)
Comment (Jannis) [reply graph/src/data/schema.rs]: I was thinking in the database schema that is generated. Say we have this entity:
```
type Foo @entity {
  bar: BigInt!
}
```
And now we add a fulltext directive called `"bar"`. Would this try to generate the `bar` column twice? Or would it generate `bar` and `_fulltext_bar` or something like that?

File: graph/src/data/schema.rs [hunk 21]
```diff
@@ -852,13 +854,18 @@ impl Schema {
...
+        // generated for entity types. The name conversions should always align with those used
+        // to generate the field names in `graph::data::schema::api::query_fields_for_type()`.
         if local_types
             .iter()
             .find(|typ| {
                 typ.fields
                     .iter()
-                    .find(|field| field.name.eq(name))
+                    .find(|field| {
+                        name == &field.name.as_str().to_camel_case()
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs]: Is the note here on line 858-859 not clear enough?
`The name conversions should always align with those used to generate the field names in graph::data::schema::api::query_fields_for_type()`.

[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs]: Oops, the reference to the API schema generation function was incorrect. It should be `graphql::schema::api::query_fields_for_type()`. 
I'll update that reference and add a similar note to `api::query_fields_for_type()`

File: graph/src/data/schema.rs [hunk 22]
```diff
@@ -862,6 +862,7 @@ impl Schema {
                     .find(|field| {
                         name == &field.name.as_str().to_camel_case()
                             || name == &field.name.to_plural().to_camel_case()
+                            || field.name.eq(name)
```
Comment (Jannis) [inline graph/src/data/schema.rs]: Ok, let's go with this for now, instead of a `_fulltext` prefix in the column names in Postgres. We can still lift this "limitation" of not being allowed to name fulltext resolvers the same as entity field names later, if we want.

File: graph/src/data/schema.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply graph/src/data/schema.rs]: Oh, sorry, I didn't see that somehow. Yes, that is good enough.

File: store/postgres/src/relational.rs [hunk 23]
```diff
@@ -408,7 +416,9 @@ impl Layout {
                 "entity_count" => entity_count
             );
         }
-        let query = FilterQuery::new(&self, collection, filter.as_ref(), order, range, block)?;
+
+        let filter_collection = FilterCollection::new(&self, collection, filter.as_ref())?;
+        let query = FilterQuery::new(&filter_collection, filter.as_ref(), order, range, block)?;
```
Comment (lutter) [inline store/postgres/src/relational.rs:421]: I don't fully understand why the `FilterCollection` needs to be built here and we therefore expose that detail. Couldn't we leave it as a detail in `FilterQuery::new`?

Comment (fordN) [reply store/postgres/src/relational.rs:421]: Yes, we definitely could.  That was meant to be an optimization to pass in a FilterCollection ref instead of the entire Layout, but upon second thought I doubt the compiled programs differ significantly if at all.  
I see your point in containing the filter building logic within FilterQuery::new(); changed back.

Comment (fordN) [reply store/postgres/src/relational.rs:421]: I moved FillterCollection new outside of FilterQuery::new() because I updated the collection property of FilterQuery to be a reference.  I couldn't make the compiler happy when creating the FilterCollection within FilterQuery: `returns a value referencing data owned by the current function`.

File: graph/src/data/schema.rs [hunk 24]
```diff
@@ -73,8 +74,201 @@ pub enum SchemaValidationError {
...
+            )),
+        }
+    }
+}
+
+impl FulltextAlgorithm {
+    pub fn as_sql(&self) -> &'static str {
+        match self {
+            Self::Rank => "ts_rank(",
+            Self::ProximityRank => "ts_rank_cd(",
```
[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs:207]: This only gets used once when building the order_by clause of a filter query, so it's a simple change to move the logic there. 
Though I like keeping `FulltextAlgorithm` manipulations contained to the trait, in this case it is _very_ Postgres specific.

[the remark this answers is not in the document]
Comment (fordN) [reply graph/src/data/schema.rs:207]: Ok, I've moved creating Postgres statements for FulltextAlgorithm into relational_queries within the `order_by()`.  
Similarly, there is a function for processing the language config into a string for use in sql; I've renamed that: `FulltextLanguage::as_sql()` -> `FulltextLanguage::as_str()` and kept it in graph::data::schema because it is generally applicable.

File: store/postgres/tests/relational.rs [hunk 25]
```diff
@@ -150,6 +150,16 @@ fn insert_entity(conn: &PgConnection, layout: &Layout, entity_type: &str, entity
...
 }
 
+fn update_entity(conn: &PgConnection, layout: &Layout, entity_type: &str, entity: Entity) {
+    let key = EntityKey {
+        subgraph_id: THINGS_SUBGRAPH_ID.clone(),
+        entity_type: entity_type.to_owned(),
+        entity_id: entity.id().unwrap(),
+    };
+    let errmsg = format!("Failed to update entity {}[{}]", entity_type, key.entity_id);
+    layout.update(&conn, &key, entity, 0).expect(&errmsg);
```
Comment (lutter) [inline store/postgres/tests/relational.rs:160]: To be truly an update, this should use block number 1, as we insert at block number 0. That will make sure that we actually have both the old and new entity in the store.

File: store/postgres/tests/relational.rs [hunk 26]
```diff
@@ -239,6 +242,53 @@ fn insert_users(conn: &PgConnection, layout: &Layout) {
...
+) {
+    let mut user = Entity::new();
+
+    user.insert("id".to_owned(), Value::String(id.to_owned()));
+    user.insert("name".to_owned(), Value::String(name.to_owned()));
+    let bin_name = Bytes::from_str(&hex::encode(name)).unwrap();
+    user.insert("bin_name".to_owned(), Value::Bytes(bin_name));
+    user.insert("email".to_owned(), Value::String(email.to_owned()));
+    user.insert(
+        "userSearch".to_owned(),
```
Comment (lutter) [inline store/postgres/tests/relational.rs:266]: This shouldn't be necessary, right?

File: store/postgres/src/relational_queries.rs [hunk 27]
```diff
@@ -1523,43 +1594,58 @@ impl QueryFragment<Pg> for FilterRange {
...
-        let collection = FilterCollection::new(layout, collection, filter)?;
-
         // Get the name of the column we order by; if there is more than one
         // table, we are querying an interface, and the order is on an attribute
         // in that interface so that all tables have a column for that. It is
         // therefore enough to just look at the first table to get the name
-        let first_table = collection
-            .first_table()
-            .expect("an entity query always contains at least one entity type/table");
+        let first_table = collection.first_table().unwrap();
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs:1618]: This is minor, but why remove the error message you get when we actually trip over this?

File: store/postgres/src/relational_queries.rs [hunk 28]
```diff
@@ -1523,43 +1594,58 @@ impl QueryFragment<Pg> for FilterRange {
 /// The parallel to `EntityQuery`.
 ///
 /// Details of how query generation for `FilterQuery` works can be found
-/// in `docs/implementation/query-prefetching.md`
+/// `https://github.com/graphprotocol/rfcs/blob/master/engineering-plans/0001-graphql-query-prefetching.md`
 #[derive(Debug, Clone)]
 pub struct FilterQuery<'a> {
-    collection: FilterCollection<'a>,
-    sort_key: SortKey,
+    collection: &'a FilterCollection<'a>,
```
Comment (lutter) [inline store/postgres/src/relational_queries.rs:1600]: For some reason, I can't add to the discussion we had before about this - but why make this a reference in the first place? There shouldn't be an issue with having that owned by `FilterQuery`

Comment (fordN) [reply store/postgres/src/relational_queries.rs:1600]: I'm remembering another reason I created the FilterCollection outside of FilterQuery::new(). I update SortKey to contain an entire Column instead of just Column.name, so now a Column reference is returned from FilterQuery::new() in the SortKey, this Column references the FilterCollection.
